### PR TITLE
[Application] dataloader for yolo @open sesame 04/07 20:08

### PR DESCRIPTION
**Pull request by baek2sm**
Add detection dataloader for yolo example.

- Set target directory, then this dataloader loads dataset
from "images" and "annotations" folders.
- Currently, It only supports "bmp" image format.
- For variable label data, it adds zero padding to label.

Signed-off-by: Seungbaek Hong <sb92.hong@samsung.com>

### CODE REVIEW

**Review by DonghakPark — APPROVED** [review]
LGTM!

File: Applications/YOLO/jni/det_dataloader.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+#include <fstream>
+#include <iostream>
+#include <nntrainer_error.h>
+#include <random>
+#include <sstream>
+
+namespace nntrainer::util {
+
+DirDataLoader::DirDataLoader(const char *directory_, unsigned int max_num_label,
+                             unsigned int c, unsigned int w, unsigned int h,
```
Comment (jijoongmoon) [inline Applications/YOLO/jni/det_dataloader.cpp:26]: It might be better to describe saying it supports bmp only for now.

**Review by jijoongmoon — APPROVED** [review]
LGTM. You need to rebase the main to resolve the CI issue for android.

File: Applications/YOLO/jni/det_dataloader.cpp (hunk 1 above)
Comment (baek2sm) [reply Applications/YOLO/jni/det_dataloader.cpp:27]: I added it, thanks.

**Comment by baek2sm** [comment]
> LGTM. You need to rebase the main to resolve the CI issue for android.

Thanks, I rebased it, but there is still a problem.
There seems to be a problem with my code. I'll try to find the cause.

File: Applications/YOLO/jni/main.cpp [hunk 2]
```diff
@@ -254,6 +253,17 @@ int main(int argc, char *argv[]) {
     model->summarize(std::cout, ML_TRAIN_SUMMARY_MODEL);
 
     model->train();
+
+    // // simple inference example
+    // model->load("./best_model.bin");
```
Comment (lhs8928) [inline Applications/YOLO/jni/main.cpp:258]: I think we need to add load function to load weight bin file before train as well

File: Applications/YOLO/jni/main.cpp [hunk 3]
```diff
@@ -254,6 +253,17 @@ int main(int argc, char *argv[]) {
     model->summarize(std::cout, ML_TRAIN_SUMMARY_MODEL);
 
     model->train();
+
+    // // simple inference example
```
Comment (lhs8928) [inline Applications/YOLO/jni/main.cpp:257]: Is there specific reason to leave this code as comment?

File: Applications/YOLO/jni/main.cpp [hunk 4]
```diff
@@ -217,7 +215,8 @@ int main(int argc, char *argv[]) {
   // create train and validation data
   std::array<UserDataType, 2> user_datas;
   try {
-    user_datas = createFakeDataGenerator(batch_size, data_size, data_split);
+    user_datas =
+      createDetDataGenerator("./train_dir/", "./valid_dir/", 5, 3, 416, 416);
```
Comment (lhs8928) [inline Applications/YOLO/jni/main.cpp:219]: Please make this values like image height/width size as constant variable

**Review by DonghakPark — COMMENTED** [review]
I think you should make ```Andrioid.mk``` file for android file or make if statement in meson.build file , for avoid android check

if you want to make ```Android.mk``` file in you can just reference ```/Application/AlexNet/jni/Android.mk``` and Application.mk file

File: Applications/YOLO/jni/main.cpp (hunk 3 above)
Comment (baek2sm) [reply Applications/YOLO/jni/main.cpp]: It was simply a reference code for inference using dummy data. I removed these codes. thanks!

File: Applications/YOLO/jni/main.cpp (hunk 4 above)
Comment (baek2sm) [reply Applications/YOLO/jni/main.cpp]: I made it using constant variables. thanks!

File: Applications/YOLO/jni/main.cpp (hunk 2 above)
Comment (baek2sm) [reply Applications/YOLO/jni/main.cpp]: I think there should be a pre-trained weights bin file to load before train.

**Comment by baek2sm** [comment]
> I think you should make `Andrioid.mk` file for android file or make if statement in meson.build file , for avoid android check
> 
> if you want to make `Android.mk` file in you can just reference `/Application/AlexNet/jni/Android.mk` and Application.mk file

Thank you so much, I will try to update it!!

File: Applications/YOLO/jni/main.cpp [hunk 5]
```diff
@@ -217,7 +215,14 @@ int main(int argc, char *argv[]) {
   // create train and validation data
   std::array<UserDataType, 2> user_datas;
   try {
-    user_datas = createFakeDataGenerator(batch_size, data_size, data_split);
+    const char *train_dir = "./train_dir/";
+    const char *valid_dir = "./valid_dir/";
+    const int max_num_label = 5;
+    const int channel = 3;
+    const int width = 416;
```
Comment (lhs8928) [inline Applications/YOLO/jni/main.cpp:222]: Please declare this variables at the beginning of this code.
width, height values are also needed in input layer.
AFAIK max_num_label is 4 isn't is?

**Comment by jijoongmoon** [comment]
you need to resolve the ci android issue : http://ci.nnstreamer.ai/nntrainer/ci/repo-workers/pr-checker/2153-202303301343320.37945890426636-78839f1307ed3e1ccfa85eae9e060523d011550e/report/build_log_2153_android_error.txt

**Review by SeoHyungjun — APPROVED** [review]
LGTM!

Can't use extensions other than .bmp when load data?
For example, can insert a file extension into a function's factor

**Comment by baek2sm** [comment]
> LGTM!
> 
> Can't use extensions other than .bmp when load data? For example, can insert a file extension into a function's factor

now it only supports bmp image files. we have to support other formats later.

**Review by lhs8928 — APPROVED** [review]
LGTM
